pkg/apis/workspaces/v1alpha1/commands.go

```diff
@@ -1,16 +1,114 @@
 	Attributes map[string]string `json:"attributes,omitempty"` // Additional command attributes
-	Name       string            `json:"name"`                 // Describes the name of the command. Should be unique per commands set.
 }
 
-type CommandAction struct {
-	Command          *string `json:"command,omitempty"`          // The actual action command-line string
-	Component        *string `json:"component,omitempty"`        // Describes component to which given action relates
-	Type             string  `json:"type"`                       // Describes action type
-	Workdir          *string `json:"workdir,omitempty"`          // Working directory where the command should be executed
-	Reference        *string `json:"reference,omitempty"`        // Working directory where the command should be executed
-	ReferenceContent *string `json:"referenceContent,omitempty"` // Working directory where the command should be executed
+type LabeledCommand struct {
+	BaseCommand `json:",inline"`
+	Label       string `json:"label,omitempty"`
+}
+
+type Command struct {
+	PolymorphicCommand `json:",inline"`
+}
+
+// +k8s:openapi-gen=true
+// +union
+type PolymorphicCommand struct {
+	// Type of workspace command
+	// +unionDiscriminator
+	// +optional
+	Type CommandType `json:"type"`
+
+	// Exec command
+	// +optional
+	Exec *ExecCommand `json:"exec,omitempty"`
+
+	// VscodeTask command
+	// +optional
+	VscodeTask *VscodeConfigurationCommand `json:"vscodeTask,omitempty"`
+
+	// VscodeLaunch command
+	// +optional
+	VscodeLaunch *VscodeConfigurationCommand `json:"vscodeLaunch,omitempty"`
+
+	// Composite command
+	// +optional
+	Composite *CompositeCommand `json:"composite,omitempty"`
+
+	// Custom command
+	// +optional
+	Custom *CustomCommand `json:"custom,omitempty"`
+}
+
+type ExecCommand struct {
+	LabeledCommand `json:",inline"`
+
+	// The actual command-line string
+	CommandLine string `json:"commandLine"`
+
+	// Describes component to which given action relates
+	Component string `json:"component,omitempty"`
+
+	// Working directory where the command should be executed
+	Workdir *string `json:"workdir,omitempty"`
+}
+
+type CompositeCommand struct {
+	LabeledCommand `json:",inline"`
+
+	// The commands that comprise this composite command
+	Commands []string `json:"commands,omitempty"`
+
+	// +optional
+	Parallel bool `json:"parallel,omitempty"`
+}
+
+// +k8s:openapi-gen=true
+// +union
+type VscodeConfigurationCommandLocation struct {
+	// Type of Vscode configuration command location
+	// +
+	// +unionDiscriminator
+	// +optional
+	LocationType string `json:"locationType"`
+
+	// Location as an absolute of relative URL
+	// +optional
+	Url string `json:"url,omitempty"`
+
+	// Embedded content of the vscode configuration file
+	// +optional
+	Inlined string `json:"inlined,omitempty"`
+}
+
+type VscodeConfigurationCommand struct {
+	BaseCommand `json:",inline"`
+	Location    VscodeConfigurationCommandLocation `json:",inline"`
```

Comment:
```suggestion
	Location    VscodeConfigurationCommandLocation `json:"location,inline"`
```

Did you miss json field name by purpose?

Reply:
Is location the same as `reference` which we have in the current Devfile model?